aioxmpp/hashes.py

```diff
@@ -0,0 +1,263 @@
+.. autofunction:: algo_of_hash
+
+.. data:: default_hash_algorithms
+
+    A set of `algo` values which consists of hash functions matching the
+    following criteria:
+
+    * They are specified as ``MUST`` or ``SHOULD`` in the supported version of
+      :xep:`300`.
+    * They are supported by :mod:`hashlib`.
+    * Only one function from each matching family is selected. If multiple
+      functions apply, ``MUST`` is preferred over ``SHOULD``.
+
+    The set thus varies based on the build of Python and possibly OpenSSL. The
+    algorithms in the set are guaranteed to return a valid hash implementation
+    when passed to :func:`~aioxmpp.misc.hash_from_algo`.
+
+    In a fully compliant build, this set consists of ``sha-256``, ``sha3-256``
+    and ``blake2b-256``.
+
+XSOs
+====
+
+.. autoclass:: Hash
+
+.. autoclass:: HashesParent()
+"""
+import hashlib
+
+import aioxmpp.xso as xso
+
+from aioxmpp.utils import namespaces
+
+namespaces.xep0300_hashes2 = "urn:xmpp:hashes:2"
+
+
+_HASH_ALGO_MAPPING = [
+    ("md2", (False, ("md2", (), {}))),
+    ("md4", (False, ("md4", (), {}))),
+    ("md5", (False, ("md5", (), {}))),
+    ("sha-1", (True, ("sha1", (), {}))),
+    ("sha-224", (True, ("sha224", (), {}))),
+    ("sha-256", (True, ("sha256", (), {}))),
+    ("sha-384", (True, ("sha384", (), {}))),
+    ("sha-512", (True, ("sha512", (), {}))),
+    ("sha3-256", (True, ("sha3_256", (), {}))),
+    ("sha3-512", (True, ("sha3_512", (), {}))),
+    ("blake2b-256", (True, ("blake2b", (), {"digest_size": 32}))),
+    ("blake2b-512", (True, ("blake2b", (), {"digest_size": 64}))),
+]
+
+
+_HASH_ALGO_MAP = dict(_HASH_ALGO_MAPPING)
+_HASH_ALGO_REVERSE_MAP = {
+    fun_name: (enabled, algo)
+    for algo, (enabled, (fun_name, fun_args, fun_kwargs)) in _HASH_ALGO_MAPPING
+    if not fun_args and not fun_kwargs
+}
+
+
+def is_algo_supported(algo):
+    try:
+        enabled, (fun_name, _, _) = _HASH_ALGO_MAP[algo]
+    except KeyError:
+        return False
+
+    return enabled and hasattr(hashlib, fun_name)
+
+
+def hash_from_algo(algo):
+    """
+    Return a :mod:`hashlib` hash given the :xep:`300` `algo`.
+
+    :param algo: The algorithm identifier as defined in :xep:`300`.
+    :type algo: :class:`str`
+    :raises NotImplementedError: if the hash algortihm is not supported by
+        :mod:`hashlib`.
+    :raises ValueError: if the hash algorithm MUST NOT be supported.
+    :return: A hash object from :mod:`hashlib` or compatible.
+
+    If the `algo` is not supported by the :mod:`hashlib` module,
+    :class:`NotImplementedError` is raised.
+    """
+
+    try:
+        enabled, (fun_name, fun_args, fun_kwargs) = _HASH_ALGO_MAP[algo]
+    except KeyError as exc:
+        raise NotImplementedError(
+            "hash algorithm {!r} unknown".format(algo)
+        ) from None
+
+    if not enabled:
+        raise ValueError(
+            "support of {} in XMPP is forbidden".format(algo)
+        )
+
+    try:
+        fun = getattr(hashlib, fun_name)
```

Comment:
Hm, I have openssl 1.1 and Python 3.5. hashlib.new("sha3_256") does not work (and hashlib.sha3_256 does not exist).

I also don’t want to try here because is_algo_supported must be reliable and I want it to work without having to create hash objects.